File: src/main/java/cn/afeibaili/mchat/socket/message/MessageManager.java
package cn.afeibaili.mchat.socket.message;

import cn.afeibaili.mchat.MChatSystem;

/**
 * 消息管理器
 *
 * @author AfeiBaili
 * @version 2026/3/3 00:33
 */

public abstract class MessageManager<Formatting, Server> {
    abstract public Formatting getFormat();

    boolean isShowInOut = true;

    abstract public void sendToMc(String message, Formatting format);

    public void sendFormattingMessageToMC(String message) {
        sendToMc(message, getFormat());
    }

    public void sendToGroup(String message) {
        MChatSystem.INSTANCE.send(new MessageType.Text(message));
    }

    public void sendLoginInOut(String message) {
        if (isShowInOut) {
            MChatSystem.INSTANCE.send(new MessageType.Text(message));
        }
    }

    public void sendHeartbeat() {
        MChatSystem.INSTANCE.send(new MessageType.Heartbeat(""));
    }


    public void parseMessage(String message) {
        String ident = message.substring(0, 4);
        String msg = message.substring(4);
        MessageType messageType = match(ident, msg);
    }

    MessageType match(String ident, String message) {
        switch (ident) {
            case "txt:": {
                MessageType.Text text = new MessageType.Text(message);
                sendToMc(text.message, getFormat());
                return text;
            }

            case "cmd:": {
                MessageType.Command command = new MessageType.Command(message);
                switch (command.message) {
                    case "help":
                    case "菜单": {
                        sendToGroup("help、菜单、close-login-message、关闭登录消息、open-login-message、开启登录消息");
                        break;
                    }
                    case "close-login-message":
                    case "关闭登录消息": {
                        isShowInOut = false;
                        sendToGroup("已关闭登录消息");
                        break;
                    }

                    case "open-login-message":
                    case "开启登录消息": {
                        isShowInOut = true;
                        sendToGroup("已开启登录消息");
                        break;
                    }
                }
            }

            case "het:": {
                return new MessageType.Heartbeat(message);
            }
            default: {
                return null;
            }
        }
    }
}

File: core/src/main/kotlin/online/afeibaili/mchat/socket/message/MessageManager.kt
package online.afeibaili.mchat.socket.message

import online.afeibaili.mchat.MChatSystem.Companion.INSTANCE


/**
 * 消息管理器
 *
 *@author AfeiBaili
 *@version 2025/11/3 16:03
 */
abstract class MessageManager<Formatting, Server>() {
    abstract var formatting: Formatting
    abstract fun sendToMC(message: String, formatting: Formatting)
    var isShowInOut: Boolean = true

    fun sendFormattingMessageToMC(message: String) {
        sendToMC(message, formatting)
    }

    fun sendToGroup(message: String) {
        INSTANCE.send(MessageType.Text(message))
    }

    fun sendLoginInOut(message: String) {
        if (isShowInOut) {
            INSTANCE.send(MessageType.Text(message))
        }
    }

    fun sendHeartbeat() {
        INSTANCE.send(MessageType.Heartbeat(""))
    }

    fun parseMessage(message: String) {
        val ident = message.take(4)
        val message: String = message.drop(4)
        match(ident, message)
    }

    private fun match(ident: String, message: String) = when (ident) {
        "txt:" -> MessageType.Text(message).apply {
            sendToMC(this.message, formatting)
        }

        "cmd:" -> MessageType.Command(message).apply {
            when (this.message) {
                "help", "菜单" -> sendToGroup(
                    """
                    菜单
                    help
                    close-login-message
                    关闭登录消息
                    open-login-message
                    开启登录消息
                """.trimIndent()
                )

                "close-login-message", "关闭登录消息" -> {
                    isShowInOut = false
                    sendToGroup("已关闭登录消息")
                }

                "open-login-message", "开启登录消息" -> {
                    isShowInOut = true
                    sendToGroup("已开启登录消息")
                }
            }
        }

        "het:" -> MessageType.Heartbeat(message)
        else -> null
    }
}